client/consensus/common/src/import_queue/buffered_link.rs

```diff
@@ -52,8 +52,10 @@ use super::BlockImportResult;
 /// Wraps around an unbounded channel from the `futures` crate. The sender implements `Link` and
 /// can be used to buffer commands, and the receiver can be used to poll said commands and transfer
 /// them to another link.
-pub fn buffered_link<B: BlockT>() -> (BufferedLinkSender<B>, BufferedLinkReceiver<B>) {
-	let (tx, rx) = tracing_unbounded("mpsc_buffered_link");
+pub fn buffered_link<B: BlockT>(
+	queue_size_warning: i64,
```

Comment:
why do we need signed integer here?

Reply:
It's explained in the comment for a struct field: to avoid underflow if, due to the lack of ordering, the counter happen to go < 0.

Reply:
Internally: yes. But public API doesn't need to be signed integer. This should have been `u32` instead that should still be plenty big for all intents and purposes. Same for `tracing_unbounded` function.

Reply:
I'm also wondering how much of a performance difference it actually makes using Relaxed ordering here.

Reply:
It's not that relaxed ordering makes sense in terms of performance, it's about not having to bother about synchronization of increments/decrements, why signed integer is used. Relaxed ordering is just a consequence of this decision, because more strong guarantees are not needed if we use the unsigned integer.

Reply:
Since in order to decrease the number you need to increase it first, `SeqCst` will do just that, you shouldn't need any extra synchronization beyond that.

Reply:
Not only this: in [`TracingUnboundedSender::unbounded_send`](https://github.com/paritytech/substrate/blob/2706277054a90d5b66fcc353b761dcfb004cdca0/client/utils/src/mpsc.rs#L145-L173) we need to make sure that the counter is incremented before the message is sent to the channel, and make sure we decrement the counter back if the call has failed. Signed integer allows not to think about such corner cases.

Reply:
What is the use case of measuring a queue size with such precision?

Reply:
Why would we put a random number when we know precisely what the number should be? In our case if queue size is at 2 we already have a problem, we know that for certain, there is no need to buffer 100 or 100_000 messages.

Reply:
I mean there is no guarantee that the queue won't grow above 2 elements, even if we are pulling them fast enough — just the pulling task might not be scheduled in time, for example. This is all probabilistic. And if we talk about clogging — that's at least hundreds of messages, and +/-1 doesn't matter here.

Reply:
Sometimes we do, we are using this in some cases for channels with explicit acknowledgement, so if something isn't acknowledged new message will not be pushed into the channel in the first place. So if we hit the threshold, something went wrong.

Reply:
Thanks for the explanation. I'll fix this in a follow-up PR.

Reply:
If you use `Acquire`/`Release` it should work: https://en.cppreference.com/w/cpp/atomic/memory_order

The compiler should add some barrier that ensures that reads/writes are not reordered.

Reply:
BTW @nazar-pc why aren't you just use a channel with a size of 0 and using try_send?

Reply:
Because I didn't see `try_send` in there, also it is usually for different access patterns. I'd expect it to still produce a warning regardless.

Reply:
Here is a PR implementing exact queue size warning (https://github.com/paritytech/substrate/pull/13117), but I'd like it to be reviewed by somebody with good understanding of concurrency, atomics, and memory order of operations. If you know who to invite for review, please invite them.